### ISSUE TITLE
Unexpected behaviour change after download worker refactor

### ISSUE THREAD

**Issue by pabl0** [iejMac/video2dataset#302]
Previously, video2dataset would create and download distribution.process_count shards at a time. So if I had 10 processes and, say, 32 threads configured, it would create ten .tar and .parquet files, and after finishing, the _stats.json file would be created (triggering my automation to upload the downloaded data elsewhere), and video2dataset would move to processing a new shard.

After the recent changes, it seems to be processing hundreds of shard at a time. Also, with process_count  4 and thread_count 16, I see some 300-400 threads running instead of the expected 64 or so. I don't think it works as designed, but if it does, how can I limit the number of shards processed at a time to make my download pipeline work smoothly?

**Comment by rom1504** [comment]
@MattUnderscoreZhang could you please take a look?

Indeed this is definitely not expected

**Comment by pabl0** [comment]
Reverting commit 83afef0 restored the old behavior for me. It seems that there is still much higher number of threads than processes_count × thread_count (haven't checked it before), but at least it creates the correct number of shards at a time.

Oh, BTW, the setting is called process_es__count, not process_count that might sound more correct (vs thread_count), for anyone grepping in the source tree. Confused me at first since the only occurrence was in the no_distributor() function.

**Comment by rom1504** [comment]
Yeah I might have merged that a bit too fast. Will look more and we can probably revert if there's no easy fix

**Comment by MattUnderscoreZhang** [comment]
I’ll take a look as soon as I get home. It’s likely a change I made to the semaphores.

**Comment by MattUnderscoreZhang** [comment]
Can you check if the threading_fix branch fixes your issue?

**Comment by MattUnderscoreZhang** [comment]
@rom1504 I created a pull request that should limit processing to the correct number of shards at a time (as tested on my local environment). However, I see the same thing that @pabl0 does, where the number of threads created is higher than expected from processes_count x thread_count, going far back in the commit history. This seems to be a bug that has existed for a long time.

**Comment by MattUnderscoreZhang** [comment]
@pabl0 I can't seem to replicate your observation. What environment are you running on, and what configs are you using? For reference, here's what I see. With five processes, I see 5 tar files being written simultaneously.
![image](https://github.com/iejMac/video2dataset/assets/19638319/b71a033a-ccdb-4edf-9e4d-2bcb62ac2295)

**Comment by MattUnderscoreZhang** [comment]
Also, I removed the print statement on the threading_fix path. Nice catch.

**Comment by pabl0** [comment]
I see. IIRC, the WebVid dataset consist of normal http links using different d/l backend than YouTube datasets. Is it possible they behave differently?
